### DISCUSSION TITLE
How to reshape a dmatrix?

### DISCUSSION THREAD

**Discussion by DennisZhangOiler** [dimforge/nalgebra#1408] [Q&A]
Hi,

I am a newbie to math and nalgebra, I have some code written in numpy and I want to translate into nalgebra, how to do np.reshape in nalgebra?

**Comment by fumagall** [comment]
Hey, 

I'm not super familiar with nalgebra and maybe there exist a simpler method but if you use the generic interface you can write:

```rust
use nalgebra::{DMatrix, Dyn, SMatrix, Const};

fn main() {
    // for dynamical sized stuff
    let mut x = DMatrix::<f64>::zeros(10, 2);
    let mut x = x.reshape_generic(Dyn(5), Dyn(4)); // works
    let mut x = x.reshape_generic(Dyn(2), Dyn(4)); // does not work but will be checked at run time
    let (n_row, n_col) = (5, 4);
    let mut x = x.reshape_generic(Dyn(n_row), Dyn(n_col)); // works
    
    // for static sized stuff
    let mut y= SMatrix::<f64, 3, 4>::zeros();
    let mut y = y.reshape_generic(Const::<6>, Const::<2>); // works
    // let mut y = y.reshape_generic(Const::<6>, Const::<1>); // does not work and you compile can already catch it
    let (n_row, n_col) = (6, 2);
    let mut y = y.reshape_generic(Const::<n_row>, Const::<n_col>); // does not work, since variables are not constant. You could use const generics.
}
```

In the `reshape_generic` you first specify the row and then the column. The `Const::<6>` tells that the size is known at compile time and has size 6, while `Dyn(6)` would say I don't know the size at compile time and it has the size 6. Since the size does not have to be known at compile time for `Dyn` you can also enter variables as well in contrast to the `Const`.
